internal/controller/replication.storage/volumegroupreplicationcontent_controller.go

```diff
@@ -18,36 +18,298 @@ package controller
+		return reconcile.Result{}, err
+	}
+
+	// Fetch VolumeGroupReplicationClass
+	vgrClass := &replicationv1alpha1.VolumeGroupReplicationClass{}
+	err = r.Client.Get(ctx, types.NamespacedName{Name: instance.Spec.VolumeGroupReplicationClassName}, vgrClass)
+	if err != nil {
+		logger.Error(err, "failed to fetch volumeGroupReplicationClass resource")
+		return reconcile.Result{}, err
+	}
+
+	// Get secrets and parameters
+	parameters := filterPrefixedParameters(replicationParameterPrefix, vgrClass.Spec.Parameters)
+	secretName := vgrClass.Spec.Parameters[prefixedGroupReplicationSecretNameKey]
+	secretNamespace := vgrClass.Spec.Parameters[prefixedGroupReplicationSecretNamespaceKey]
+
+	// Get the VolumeGroupReplication resource
+	if instance.Spec.VolumeGroupReplicationRef == nil {
+		logger.Info("waiting for owner VolumeGroupReplication to update the owner ref")
+		return reconcile.Result{}, nil
+	}
+	vgrObj := &replicationv1alpha1.VolumeGroupReplication{}
+	namespacedObj := types.NamespacedName{Namespace: instance.Spec.VolumeGroupReplicationRef.Namespace,
+		Name: instance.Spec.VolumeGroupReplicationRef.Name}
+	err = r.Client.Get(ctx, namespacedObj, vgrObj)
+	if err != nil {
+		logger.Error(err, "failed to get owner VolumeGroupReplication")
+		return reconcile.Result{}, err
+	}
+
+	// Check if object is being deleted
+	if instance.GetDeletionTimestamp().IsZero() {
+		// add vgr as the owner annotation for vgrContent
+		err = r.annotateVolumeGroupReplicationWithVGRContentOwner(ctx, logger, req.Name, vgrObj)
+		if err != nil {
+			logger.Error(err, "Failed to annotate VolumeGroupReplication owner")
+			return ctrl.Result{}, err
+		}
+		// add vgr finalizer to vgrContent
+		if err = AddFinalizerToVGRContent(r.Client, logger, instance, vgrReplicationFinalizer); err != nil {
+			logger.Error(err, "failed to add VolumeGroupReplicationContent finalizer")
+			return reconcile.Result{}, err
+		}
+	} else {
+		// Check if the owner VGR is marked for deletion, only then remove the finalizer from VGRContent resource
+		if vgrObj.GetDeletionTimestamp().IsZero() {
+			logger.Info("cannot delete VolumeGroupReplicationContent resource, until owner VolumeGroupReplication instance is deleted")
+			return reconcile.Result{}, nil
+		} else if !slices.Contains(instance.ObjectMeta.Finalizers, volumeReplicationFinalizer) {
+			// Delete the volume group, if exists
+			if instance.Spec.VolumeGroupReplicationHandle != "" {
+				// Fetch the volume group using the group id
+				tmpVG, err := volumeGroupClient.ControllerGetVolumeGroup(instance.Spec.VolumeGroupReplicationHandle, secretName, secretNamespace)
+				if err != nil {
+					if status.Code(err) == codes.NotFound {
+						logger.Info("failed to get volume group, the volume group has already been deleted or doesn't exists")
+					} else {
+						logger.Error(err, "failed to get the volume group to be deleted", "Volume Group", instance.Spec.VolumeGroupReplicationHandle)
+						return reconcile.Result{}, err
+					}
```

Comment:
why do we need to do get?

We should issue modify and delete directly

Reply:
Earlier, we had the bug for DeleteVolumeGroup RPC for not being idempetent so added this and later on forgot to remove it.
Fixed it now.